### PR TITLE
When componentRequirements() contains kind BINDING don't generate component create().

### PR DESCRIPTION
**Pull request by naturalwarren**
When a user specifies a component builder will provide an instance for a binding a component cannot instantiate all its requirements. As such, we should not generate component.create().

Discussed in #584.

### CODE REVIEW

**Comment by naturalwarren** [comment]
I signed it!

File: compiler/src/main/java/dagger/internal/codegen/Util.java [hunk 1]
```diff
@@ -43,10 +43,16 @@
...
 final class Util {
   /**
-   * Returns true if the passed {@link TypeElement} requires a passed instance in order to be used
+   * Returns true if the passed {@link ComponentRequirement} requires a passed instance in order to be used
    * within a component.
    */
-  static boolean requiresAPassedInstance(Elements elements, Types types, TypeElement typeElement) {
+  static boolean requiresAPassedInstance(Elements elements, Types types, ComponentRequirement componentRequirement) {
+    if (componentRequirement.kind() == ComponentRequirement.Kind.BINDING) {
+      /* A user has explicitly defined in their component builder they will provide an instance. */
```
Comment (ronshapiro) [inline compiler/src/main/java/dagger/internal/codegen/Util.java]: Can you make this a `// single line comment` instead?

File: compiler/src/main/java/dagger/internal/codegen/Util.java [hunk 2]
```diff
@@ -43,10 +43,16 @@
  */
 final class Util {
   /**
-   * Returns true if the passed {@link TypeElement} requires a passed instance in order to be used
+   * Returns true if the passed {@link ComponentRequirement} requires a passed instance in order to be used
```
Comment (ronshapiro) [inline compiler/src/main/java/dagger/internal/codegen/Util.java]: Can you reformat this javadoc to not exceed 100 characters?

File: compiler/src/test/java/dagger/internal/codegen/ComponentBuilderTest.java [hunk 3]
```diff
@@ -421,6 +421,70 @@ public void testBuilderMissingBuildMethodFails() {
...
+                    "test.SimpleComponent",
+                    "package test;",
+                    "",
+                    "import dagger.BindsInstance;",
+                    "import dagger.Component;",
+                    "import javax.inject.Provider;",
+                    "",
+                    "@Component",
+                    "interface SimpleComponent {",
+                    "  @Component.Builder",
```
Comment (ronshapiro) [inline compiler/src/test/java/dagger/internal/codegen/ComponentBuilderTest.java]: Can you add a component method for `Object object();`?

File: compiler/src/main/java/dagger/internal/codegen/Util.java (hunk 2 above)
Comment (naturalwarren) [inline compiler/src/main/java/dagger/internal/codegen/Util.java]: Done.

File: compiler/src/main/java/dagger/internal/codegen/Util.java (hunk 1 above)
Comment (naturalwarren) [inline compiler/src/main/java/dagger/internal/codegen/Util.java]: Done.

File: compiler/src/test/java/dagger/internal/codegen/ComponentBuilderTest.java (hunk 3 above)
Comment (naturalwarren) [inline compiler/src/test/java/dagger/internal/codegen/ComponentBuilderTest.java]: Done.

**Comment by naturalwarren** [comment]
@ronshapiro Thanks for reviewing this. I've improved the styling of the JavaDocs and added a component method in the test case as you suggested.

File: compiler/src/test/java/dagger/internal/codegen/ComponentBuilderTest.java [hunk 4]
```diff
@@ -421,6 +421,87 @@ public void testBuilderMissingBuildMethodFails() {
...
+                    "package test;",
+                    "",
+                    "import dagger.BindsInstance;",
+                    "import dagger.Component;",
+                    "import javax.inject.Provider;",
+                    "",
+                    "@Component",
+                    "interface SimpleComponent {",
+                    "  Object object();",
+                    "  @Component.Builder",
```
Comment (ronshapiro) [inline compiler/src/test/java/dagger/internal/codegen/ComponentBuilderTest.java]: nit: can you add a line before `@Component.Builder`?

File: compiler/src/test/java/dagger/internal/codegen/ComponentBuilderTest.java [hunk 5]
```diff
@@ -421,6 +421,87 @@ public void testBuilderMissingBuildMethodFails() {
...
+                    "",
+                    "  @SuppressWarnings(\"unchecked\")",
+                    "  private void initialize(final Builder builder) {",
+                    "",
+                    "    this.objectProvider = InstanceFactory.create(builder.object);",
+                    "  }",
+                    "",
+                    "  @Override",
+                    "  public Object object() {",
+                    "  return objectProvider.get();",
```
Comment (ronshapiro) [inline compiler/src/test/java/dagger/internal/codegen/ComponentBuilderTest.java]: nit: indent the return

File: compiler/src/test/java/dagger/internal/codegen/ComponentBuilderTest.java [hunk 6]
```diff
@@ -421,6 +421,87 @@ public void testBuilderMissingBuildMethodFails() {
...
+                    "    initialize(builder);",
+                    "  }",
+                    "",
+                    "  public static SimpleComponent.Builder builder() {",
+                    "    return new Builder();",
+                    "  }",
+                    "",
+                    "  @SuppressWarnings(\"unchecked\")",
+                    "  private void initialize(final Builder builder) {",
+                    "",
```
Comment (ronshapiro) [inline compiler/src/test/java/dagger/internal/codegen/ComponentBuilderTest.java]: nit: remove

**Pull request closed without merging**